File: backend/app/schemas/purchasing.py
from datetime import datetime

from pydantic import BaseModel, Field


class PurchaseRequestCreate(BaseModel):
    material: str
    color: str
    thickness: float
    requested_area_m2: float = Field(gt=0)
    note: str | None = None
    # История цен и сроков поставщика — оба необязательны, могут быть
    # согласованы позже создания заявки (см. PurchaseRequestUpdate).
    supplier: str | None = None
    price_per_m2: float | None = Field(default=None, gt=0)


class PurchaseRequestUpdate(BaseModel):
    """Заявка уже создана, но цена/поставщик согласованы позже — например,
    после переговоров. Материал/цвет/толщина/объём неизменны."""

    supplier: str | None = None
    price_per_m2: float | None = Field(default=None, gt=0)


class PurchaseRequestShopFloorCreate(BaseModel):
    """Заявка "с цеха" (раздел про замену "Заказов покупателей") — с
    "Выдачи участку" по нехватке остатка под конкретную строку задания.
    Без supplier/price_per_m2 — их выбирает снабженец позже на "Закупках",
    как и у обычной заявки после создания."""

    material: str
    color: str
    thickness: float
    requested_area_m2: float = Field(gt=0)
    note: str | None = None


class PurchaseRequestFulfillRequest(BaseModel):
    """Привязка заявки к конкретной приёмке по УПД (раздел про ускорение
    приёмки) — закрывает заявку и запоминает, каким УПД она закрыта."""

    upd_number: str


class PurchaseRequestOut(BaseModel):
    id: int
    material: str
    color: str
    thickness: float
    requested_area_m2: float
    current_stock_m2: float
    note: str | None
    status: str
    origin: str
    linked_upd_number: str | None
    created_by: int
    created_at: datetime
    closed_at: datetime | None
    supplier: str | None
    price_per_m2: float | None

File: frontend/src/api/purchasing.ts
import { apiClient } from "./client";

export interface PurchaseRequest {
  id: number;
  material: string;
  color: string;
  thickness: number;
  requested_area_m2: number;
  current_stock_m2: number;
  note: string | null;
  status: string;
  origin: string;
  linked_upd_number: string | null;
  created_by: number;
  created_at: string;
  closed_at: string | null;
  supplier: string | null;
  price_per_m2: number | null;
}

export interface PurchaseRequestCreate {
  material: string;
  color: string;
  thickness: number;
  requested_area_m2: number;
  note?: string;
  supplier?: string;
  price_per_m2?: number;
}

export interface PurchaseRequestShopFloorCreate {
  material: string;
  color: string;
  thickness: number;
  requested_area_m2: number;
  note?: string;
}

export interface PurchaseRequestUpdate {
  supplier?: string;
  price_per_m2?: number;
}

export async function listPurchaseRequests(statusFilter?: string): Promise<PurchaseRequest[]> {
  const { data } = await apiClient.get<PurchaseRequest[]>("/purchase-requests", {
    params: statusFilter ? { status_filter: statusFilter } : undefined,
  });
  return data;
}

export async function createPurchaseRequest(payload: PurchaseRequestCreate): Promise<PurchaseRequest> {
  const { data } = await apiClient.post<PurchaseRequest>("/purchase-requests", payload);
  return data;
}

// Заявка "с цеха" — кнопка "Подать заявку на закупку" на "Выдаче участку"
// при нехватке остатка под строку задания.
export async function createShopFloorPurchaseRequest(payload: PurchaseRequestShopFloorCreate): Promise<PurchaseRequest> {
  const { data } = await apiClient.post<PurchaseRequest>("/purchase-requests/shop-floor", payload);
  return data;
}

export async function updatePurchaseRequest(id: number, payload: PurchaseRequestUpdate): Promise<PurchaseRequest> {
  const { data } = await apiClient.patch<PurchaseRequest>(`/purchase-requests/${id}`, payload);
  return data;
}

export async function closePurchaseRequest(id: number): Promise<PurchaseRequest> {
  const { data } = await apiClient.post<PurchaseRequest>(`/purchase-requests/${id}/close`);
  return data;
}

// Привязка заявки к конкретной приёмке по УПД (раздел про ускорение
// приёмки) — закрывает заявку и запоминает, каким УПД она была закрыта.
export async function fulfillPurchaseRequest(id: number, updNumber: string): Promise<PurchaseRequest> {
  const { data } = await apiClient.post<PurchaseRequest>(`/purchase-requests/${id}/fulfill`, { upd_number: updNumber });
  return data;
}
